# Seven Days of Code - Day 1

### Data Cleaning

Day one of Seven Days of Code project
<p>More about the project <a href="https://7daysofcode.io/matricula/data-science">
here</a>.</p>

<p>Reference <a href="https://www.tableau.com/learn/articles/what-is-data-cleaning">
here</a>.</p>


<p>Data Source <a href="https://www12.senado.leg.br/transparencia/dados-abertos-transparencia/dados-abertos-ceaps?utm_source=ActiveCampaign&utm_medium=email&utm_content=%237DaysOfCode+-+Ci%C3%AAncia+de+Dados+1%2F7%3A+Data+Cleaning+and+Preparation&utm_campaign=%5BAlura+%237Days+Of+Code%5D%28Java%29+Dia+1%2F7%3A+Consumir+uma+API+de+filmes" >here</a> </p>

In [43]:
#Import the libraries
import pandas as pd

In [44]:
#Import the data
#Senators declared expenses in 2021
data = pd.read_csv('despesa_ceaps_2021.csv',encoding='ansi',sep=';')

In [45]:
data.head()

ANO  MES       SENADOR                                       TIPO_DESPESA  \
0  2021    1  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
1  2021    1  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
2  2021    1  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
3  2021    1  ACIR GURGACZ                Divulgação da atividade parlamentar   
4  2021    1  ACIR GURGACZ                Divulgação da atividade parlamentar   

             CNPJ_CPF                                         FORNECEDOR  \
0  05.914.650/0001-66                                           ENERGISA   
1      062.135.728-64  FERNANDO WALDEIR PACINI e ANA LUCIA DA SILVA S...   
2      004.948.028-63                      GILBERTO PISELO DO NASCIMENTO   
3  26.320.603/0001-64                                      INFORMANAHORA   
4  13.659.201/0001-47                    LINHA PURPURA FOTO E VIDEO LTDA   

    DOCUMENTO        DATA                                       DETALHAMENTO  \
0  23.489.627  21/01/2021  Pagamento de energia elétrica para uso do escr...   
1      jan/21  05/01/2021  Despesa com pagamento de aluguel de imóvel par...   
2      001/21  06/01/2021  Despesa com aluguel de imóvel para uso do escr...   
3          10  25/01/2021                Divulgação da atividade parlamentar   
4          44  07/01/2021                Divulgação da atividade parlamentar   

  VALOR_REEMBOLSADO  COD_DOCUMENTO  
0             75,29        2158003  
1              1000        2157367  
2              6000        2156383  
3              1500        2154509  
4              6000        2154507

In [46]:
#Check dataframe shape
data.shape

(16827, 11)

### Step 1 - Remove duplicates

In [47]:
#Remove duplicates
data.drop_duplicates(inplace=True)

In [48]:
data.shape

(16827, 11)

It seems that there wasn't any duplicates in this dataset

### Step 2 - Check null values

In [49]:
#Check null values
data.isnull()

ANO    MES  SENADOR  TIPO_DESPESA  CNPJ_CPF  FORNECEDOR  DOCUMENTO  \
0      False  False    False         False     False       False      False   
1      False  False    False         False     False       False      False   
2      False  False    False         False     False       False      False   
3      False  False    False         False     False       False      False   
4      False  False    False         False     False       False      False   
...      ...    ...      ...           ...       ...         ...        ...   
16822  False  False    False         False     False       False      False   
16823  False  False    False         False     False       False      False   
16824  False  False    False         False     False       False      False   
16825  False  False    False         False     False       False      False   
16826  False  False    False         False     False       False      False   

        DATA  DETALHAMENTO  VALOR_REEMBOLSADO  COD_DOCUMENTO  
0      False         False              False          False  
1      False         False              False          False  
2      False         False              False          False  
3      False         False              False          False  
4      False         False              False          False  
...      ...           ...                ...            ...  
16822  False         False              False          False  
16823  False         False              False          False  
16824  False         False              False          False  
16825  False         False              False          False  
16826  False         False              False          False  

[16827 rows x 11 columns]

In [50]:
#Check only if there are null values for "VALOR_REEMBOLSADO"
valor_series = pd.isnull(data['VALOR_REEMBOLSADO'])

In [51]:
data[valor_series]

Empty DataFrame
Columns: [ANO, MES, SENADOR, TIPO_DESPESA, CNPJ_CPF, FORNECEDOR, DOCUMENTO, DATA, DETALHAMENTO, VALOR_REEMBOLSADO, COD_DOCUMENTO]
Index: []

There aren't null values for 'VALOR_REEMBOLSADO'

In [52]:
#Check only if there are null values for "DETALHAMENTO"
detal_series = pd.isnull(data['DETALHAMENTO'])

In [53]:
data[detal_series]

ANO  MES            SENADOR  \
175    2021    1  ALESSANDRO VIEIRA   
176    2021    1  ALESSANDRO VIEIRA   
177    2021    1  ALESSANDRO VIEIRA   
178    2021    1  ALESSANDRO VIEIRA   
179    2021    1  ALESSANDRO VIEIRA   
...     ...  ...                ...   
16816  2021   12   ZEQUINHA MARINHO   
16817  2021   12   ZEQUINHA MARINHO   
16818  2021   12   ZEQUINHA MARINHO   
16819  2021   12   ZEQUINHA MARINHO   
16820  2021   12   ZEQUINHA MARINHO   

                                            TIPO_DESPESA            CNPJ_CPF  \
175    Aluguel de imóveis para escritório político, c...  28.839.939/0001-63   
176    Aluguel de imóveis para escritório político, c...  06.028.059/0001-74   
177    Aluguel de imóveis para escritório político, c...  40.432.544/0107-03   
178    Aluguel de imóveis para escritório político, c...  13.017.462/0001-63   
179    Aquisição de material de consumo para uso no e...  11.088.468/0001-04   
...                                                  ...                 ...   
16816  Locomoção, hospedagem, alimentação, combustíve...  84.146.638/0001-25   
16817  Locomoção, hospedagem, alimentação, combustíve...  02.280.133/0066-30   
16818  Locomoção, hospedagem, alimentação, combustíve...  27.838.829/0001-14   
16819  Locomoção, hospedagem, alimentação, combustíve...  63.814.339/0001-41   
16820  Locomoção, hospedagem, alimentação, combustíve...  37.808.513/0001-33   

                                              FORNECEDOR    DOCUMENTO  \
175    CASA DESIGN SERVICOS DE APOIO ADMINISTRATIVO LTDA          853   
176                      CASA SHOPPING DESING LTDA - EPP  71943790582   
177                                      CLARO S.A / NET     2,10E+12   
178                                             ENERGISA   31.280.321   
179                                KAFLEX COMERCIAL LTDA     2,02E+14   
...                                                  ...          ...   
16816                          POSTO CAVALO DE AÇO LTDA.        74036   
16817                                   POSTO ICCAR LTDA        29485   
16818                             POSTO JOAO NETO EIRELI       325321   
16819                           POSTO SÃO DOMINGOS LTDA.       855410   
16820                 ROTA AUTO POSTO COMBUSTIVEL EIRELI        22422   

             DATA DETALHAMENTO VALOR_REEMBOLSADO  COD_DOCUMENTO  
175    15/01/2021          NaN               500        2155145  
176    09/03/2021          NaN              3300        2156928  
177    01/01/2021          NaN            247,96        2156078  
178    13/01/2021          NaN             293,8        2156076  
179    01/02/2021          NaN               600        2156075  
...           ...          ...               ...            ...  
16816  10/12/2021          NaN            200,03        2172714  
16817  01/12/2021          NaN            210,02        2171582  
16818  01/12/2021          NaN            116,04        2171588  
16819  18/12/2021          NaN               250        2173255  
16820  02/12/2021          NaN             35,07        2171589  

[7626 rows x 11 columns]

There are 7626 rows without any description for 'DETALHAMENTO' </p>
I'll replace it by "Detalhamento Faltante"

In [54]:
#Replace 'DETALHAMENTO' null values
data['DETALHAMENTO'].fillna('Detalhamento Faltante',inplace=True)

In [55]:
data.loc[data['DETALHAMENTO'] == 'Detalhamento Faltante']

ANO  MES            SENADOR  \
175    2021    1  ALESSANDRO VIEIRA   
176    2021    1  ALESSANDRO VIEIRA   
177    2021    1  ALESSANDRO VIEIRA   
178    2021    1  ALESSANDRO VIEIRA   
179    2021    1  ALESSANDRO VIEIRA   
...     ...  ...                ...   
16816  2021   12   ZEQUINHA MARINHO   
16817  2021   12   ZEQUINHA MARINHO   
16818  2021   12   ZEQUINHA MARINHO   
16819  2021   12   ZEQUINHA MARINHO   
16820  2021   12   ZEQUINHA MARINHO   

                                            TIPO_DESPESA            CNPJ_CPF  \
175    Aluguel de imóveis para escritório político, c...  28.839.939/0001-63   
176    Aluguel de imóveis para escritório político, c...  06.028.059/0001-74   
177    Aluguel de imóveis para escritório político, c...  40.432.544/0107-03   
178    Aluguel de imóveis para escritório político, c...  13.017.462/0001-63   
179    Aquisição de material de consumo para uso no e...  11.088.468/0001-04   
...                                                  ...                 ...   
16816  Locomoção, hospedagem, alimentação, combustíve...  84.146.638/0001-25   
16817  Locomoção, hospedagem, alimentação, combustíve...  02.280.133/0066-30   
16818  Locomoção, hospedagem, alimentação, combustíve...  27.838.829/0001-14   
16819  Locomoção, hospedagem, alimentação, combustíve...  63.814.339/0001-41   
16820  Locomoção, hospedagem, alimentação, combustíve...  37.808.513/0001-33   

                                              FORNECEDOR    DOCUMENTO  \
175    CASA DESIGN SERVICOS DE APOIO ADMINISTRATIVO LTDA          853   
176                      CASA SHOPPING DESING LTDA - EPP  71943790582   
177                                      CLARO S.A / NET     2,10E+12   
178                                             ENERGISA   31.280.321   
179                                KAFLEX COMERCIAL LTDA     2,02E+14   
...                                                  ...          ...   
16816                          POSTO CAVALO DE AÇO LTDA.        74036   
16817                                   POSTO ICCAR LTDA        29485   
16818                             POSTO JOAO NETO EIRELI       325321   
16819                           POSTO SÃO DOMINGOS LTDA.       855410   
16820                 ROTA AUTO POSTO COMBUSTIVEL EIRELI        22422   

             DATA           DETALHAMENTO VALOR_REEMBOLSADO  COD_DOCUMENTO  
175    15/01/2021  Detalhamento Faltante               500        2155145  
176    09/03/2021  Detalhamento Faltante              3300        2156928  
177    01/01/2021  Detalhamento Faltante            247,96        2156078  
178    13/01/2021  Detalhamento Faltante             293,8        2156076  
179    01/02/2021  Detalhamento Faltante               600        2156075  
...           ...                    ...               ...            ...  
16816  10/12/2021  Detalhamento Faltante            200,03        2172714  
16817  01/12/2021  Detalhamento Faltante            210,02        2171582  
16818  01/12/2021  Detalhamento Faltante            116,04        2171588  
16819  18/12/2021  Detalhamento Faltante               250        2173255  
16820  02/12/2021  Detalhamento Faltante             35,07        2171589  

[7626 rows x 11 columns]

### Step 2 - check NA values

In [56]:
#Check NA values
data.isna()

ANO    MES  SENADOR  TIPO_DESPESA  CNPJ_CPF  FORNECEDOR  DOCUMENTO  \
0      False  False    False         False     False       False      False   
1      False  False    False         False     False       False      False   
2      False  False    False         False     False       False      False   
3      False  False    False         False     False       False      False   
4      False  False    False         False     False       False      False   
...      ...    ...      ...           ...       ...         ...        ...   
16822  False  False    False         False     False       False      False   
16823  False  False    False         False     False       False      False   
16824  False  False    False         False     False       False      False   
16825  False  False    False         False     False       False      False   
16826  False  False    False         False     False       False      False   

        DATA  DETALHAMENTO  VALOR_REEMBOLSADO  COD_DOCUMENTO  
0      False         False              False          False  
1      False         False              False          False  
2      False         False              False          False  
3      False         False              False          False  
4      False         False              False          False  
...      ...           ...                ...            ...  
16822  False         False              False          False  
16823  False         False              False          False  
16824  False         False              False          False  
16825  False         False              False          False  
16826  False         False              False          False  

[16827 rows x 11 columns]

In [57]:
#Check NA for'VALOR_REEMBOLSADO'
valorna_series = pd.isna(data['VALOR_REEMBOLSADO'])

In [58]:
data[valorna_series]

Empty DataFrame
Columns: [ANO, MES, SENADOR, TIPO_DESPESA, CNPJ_CPF, FORNECEDOR, DOCUMENTO, DATA, DETALHAMENTO, VALOR_REEMBOLSADO, COD_DOCUMENTO]
Index: []

In [59]:
#Check NA for 'SENADOR'
senador_series = pd.isna(data['SENADOR'])
data[senador_series]

Empty DataFrame
Columns: [ANO, MES, SENADOR, TIPO_DESPESA, CNPJ_CPF, FORNECEDOR, DOCUMENTO, DATA, DETALHAMENTO, VALOR_REEMBOLSADO, COD_DOCUMENTO]
Index: []

In [60]:
#Check NA for 'DATA'
data_series = pd.isna(data['DATA'])
data[data_series]

Empty DataFrame
Columns: [ANO, MES, SENADOR, TIPO_DESPESA, CNPJ_CPF, FORNECEDOR, DOCUMENTO, DATA, DETALHAMENTO, VALOR_REEMBOLSADO, COD_DOCUMENTO]
Index: []

In [61]:
#Check NA for 'TIPO_DESPESA	'
tipo_series = pd.isna(data['TIPO_DESPESA'])
data[tipo_series]

Empty DataFrame
Columns: [ANO, MES, SENADOR, TIPO_DESPESA, CNPJ_CPF, FORNECEDOR, DOCUMENTO, DATA, DETALHAMENTO, VALOR_REEMBOLSADO, COD_DOCUMENTO]
Index: []

For the most important variables I want to look, there aren't NA values

### Step 4 - Remove unnecessary columns

In [62]:
#Remove unnecessary columns
data.drop(columns = ['DOCUMENTO','COD_DOCUMENTO','CNPJ_CPF'],axis=1,inplace=True)

In [63]:
data.head(2)

ANO  MES       SENADOR                                       TIPO_DESPESA  \
0  2021    1  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
1  2021    1  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   

                                          FORNECEDOR        DATA  \
0                                           ENERGISA  21/01/2021   
1  FERNANDO WALDEIR PACINI e ANA LUCIA DA SILVA S...  05/01/2021   

                                        DETALHAMENTO VALOR_REEMBOLSADO  
0  Pagamento de energia elétrica para uso do escr...             75,29  
1  Despesa com pagamento de aluguel de imóvel par...              1000

### Step 5 - make sure all values in 'VALOR_REEMBOLSADO' are numeric

In [64]:
#Make sure all values in 'VALOR_REEMBOLSADO' are numeric
valor_series_num = pd.Series(data['VALOR_REEMBOLSADO'])
valor_series_num.str.isnumeric()

0        False
1         True
2         True
3         True
4         True
         ...  
16822    False
16823    False
16824    False
16825    False
16826    False
Name: VALOR_REEMBOLSADO, Length: 16827, dtype: bool

In [65]:
#See one example of non-numeric
data['VALOR_REEMBOLSADO'].iloc[16822]

'1715,45'

It might have something to do with the decimal separator

In [66]:
#When I try to calculate a simple sum with 'VALOR_REEMBOLSADO' in line 16822. it returns an error
data['VALOR_REEMBOLSADO'].iloc[16822] + 1

TypeError: can only concatenate str (not "int") to str

In [67]:
#However, when I try to concatenate a string, it works
data['VALOR_REEMBOLSADO'].iloc[16822] + 'A'

'1715,45A'

In [68]:
#Replace "," to "." as decimal separator using list comprehension
data['VALOR_REEMBOLSADO'] = [x.replace(',','.') for x in data['VALOR_REEMBOLSADO']]

In [69]:
#Make sure 'VALOR_REEMBOLSADO' is a float
data['VALOR_REEMBOLSADO'] = data['VALOR_REEMBOLSADO'].astype(float)

In [70]:
#Check if the change was applied
data['VALOR_REEMBOLSADO'].iloc[16822]

1715.45

In [71]:
#Now, the calculation works
data['VALOR_REEMBOLSADO'].iloc[16822] + 1

1716.45

### Step 6 - confirm if Data indeed has a correct data format

In [72]:
data['DATA']

0        21/01/2021
1        05/01/2021
2        06/01/2021
3        25/01/2021
4        07/01/2021
            ...    
16822    01/12/2021
16823    15/12/2021
16824    17/12/2021
16825    27/12/2021
16826    29/12/2021
Name: DATA, Length: 16827, dtype: object

In [73]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16827 entries, 0 to 16826
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ANO                16827 non-null  int64  
 1   MES                16827 non-null  int64  
 2   SENADOR            16827 non-null  object 
 3   TIPO_DESPESA       16827 non-null  object 
 4   FORNECEDOR         16827 non-null  object 
 5   DATA               16827 non-null  object 
 6   DETALHAMENTO       16827 non-null  object 
 7   VALOR_REEMBOLSADO  16827 non-null  float64
dtypes: float64(1), int64(2), object(5)
memory usage: 1.2+ MB


In [76]:
# convert the 'DATA' column to datetime format
data['DATA'] = pd.to_datetime(data['DATA'])

In [77]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16827 entries, 0 to 16826
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   ANO                16827 non-null  int64         
 1   MES                16827 non-null  int64         
 2   SENADOR            16827 non-null  object        
 3   TIPO_DESPESA       16827 non-null  object        
 4   FORNECEDOR         16827 non-null  object        
 5   DATA               16827 non-null  datetime64[ns]
 6   DETALHAMENTO       16827 non-null  object        
 7   VALOR_REEMBOLSADO  16827 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 1.2+ MB


### End